experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+		channelAttributes: Readonly<IChannelAttributes>,
+	): Promise<IChannel> {
+		const sharedSomething = new SharedSomething(
+			runtime,
+			id,
+			this.options /* uh oh: channelAttributes and this.options might disagree on snapshot format! */,
+		);
+		await sharedSomething.load(services);
+		return sharedSomething;
+	}
+
+	public create(runtime: IFluidDataStoreRuntime, id: string): IChannel {
+		const sharedSomething = new SharedSomething(document, id, this.options);
+		sharedSomething.initializeLocal();
+		return sharedSomething;
+	}
+}
+```
+
+If the channel's attributes clearly reflect its parameters (e.g. the snapshot format being used),
+then the resulting factory sometimes produces DDSes which don't match the requested format!
+
+There are essentially two ways to resolve this issue:
+
+1. Make `SharedSomethingFactory` have a parameterless constructor, and pass parameters all the way through the `create` flow (i.e. through `createChannel`, factory lookup, and factory's `create` method).
+2. Keep parameters in `SharedSomethingFactory`'s constructor, but change their semantics such that they only necessarily apply to newly created documents.
+
+Regardless of which option we pick, enough information to reify the DDS's parameters needs to be persisted in the attributes blob.
+
+In order to support data objects which can create multiple types of the same shared object with different parameters, option 2 would need to be accompanied by changes to make DDS registration take in named entries in `@fluidframework/aqueduct`, rather than assume the "key" for a DDS registry entry is that DDS's factory's `type` field.
+Option 2 would also necessarily force us away from the static `create` method commonly found on DDSes.
+
+```typescript
+// Container author code, from the perspective of a container author
+// Option 1:
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [SharedSomething.getFactory()]);
+
+// later, they make a SharedSomething:
+const something = SharedSomething.create({ formatVersion: 5 });
+
+// Option 2:
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	["v4Something", SharedSomething.getFactory({ formatVersion: 4 })],
+	["v5Something", SharedSomething.getFactory({ formatVersion: 5 })],
+	// For API back-compat, we could also accept just SharedSomething factories
+	// and behind the scenes convert it to
+	// [factory.type, factory]
+	SharedSomething.getFactory(/* whatever the default options are */),
+]);
+
+// later, they make a v5 SharedSomething:
+const something = myDataObject.runtime.createChannel("v5Something");
+```
+
+The difference between #1 and #2 is then largely whether an author declares parameters for their DDS at _creation time_, or at _data object declaration time_.
+
+Option #1 is roughly prototyped in [this commit](https://github.com/microsoft/FluidFramework/commit/02e34ce16ac543727159f8009244832c572ba442).
+In general, I find option #2 the better choice.
```

Comment:
i don't really like either of these options. i would prefer something that looks more like dependency injection, but i don't have a super clear picture of what that would look like at the moment, but i can think of some properties that it would have:
- The data store would control which dependencies are injected into the dds
- The injection would need to happen before data is loaded, and likely be part of the channel initialization process, and happen at the same point for both load and create

I also have worries about how the listed approached work on the upgrade path, it seems like each registry will need multiple entries per config, one to load the old version and write the new version, and one to load and write the new version.